### ISSUE TITLE
GCD semaphores might be unsafe when used in signal handlers

### ISSUE THREAD

**Issue by henninglive** [adrienverge/openfortivpn#105]
Openfortivpn [currently uses](https://github.com/adrienverge/openfortivpn/blob/master/src/io.c#L33-L45) GCD (Grand Central Dispatch) semaphores on OS-X. This is done to work around the fact the that OS-X doesn’t support [unnamed posix semaphores](http://man7.org/linux/man-pages/man7/sem_overview.7.html). I ran into the same problem while working on another project and while looking for workarounds I found multiple solutions suggesting substituting posix semaphores with GCD semaphores, but I was not able to find any documentation specifying if GCD semaphores are async-signal safe.

I did some further dinging and found the [source](https://github.com/apple/swift-corelibs-libdispatch) for GCD. As you can see [here](https://github.com/apple/swift-corelibs-libdispatch/blob/d04a0dff9f0ecf327d9fc7626f3013be4d9353f5/src/semaphore.c#L239-L251) and [here](https://github.com/apple/swift-corelibs-libdispatch/blob/d04a0dff9f0ecf327d9fc7626f3013be4d9353f5/src/semaphore.c#L220-L237), GCD uses `mach/semaphore.h` internally. Unfortunately, the [documentation](https://developer.apple.com/library/content/documentation/Darwin/Conceptual/KernelProgramming/synchronization/synchronization.html) for `mach/semaphore.h` suggest that they are not async-signal safe.
>Semaphores can be used any place where mutexes can occur. This precludes their use in interrupt handlers or within the context of the scheduler, and makes it strongly discouraged in the VM system. 

If this is correct, it is properly a bad idea to use GCD semaphores.  The currently solution might work most of the time, but specific timing might cause it to fail. [Here](https://docs.oracle.com/cd/E19455-01/806-5257/gen-26/index.html) is more info on Async-Signal Safety.

[Here](https://bugs.openjdk.java.net/browse/JDK-8129814) is another discussion about the same problem in OpenJDK. They suggest using [BSD kqueue/kevent](https://www.freebsd.org/cgi/man.cgi?kqueue).

In the other project I was working on we ended up using pipes, similar to [this](http://ck-hack.blogspot.no/2013/09/unnamed-semaphores-and-pososx.html). This should work on all posix compliant systems, unlike unnamed semaphores which are an optional part of the posix standard.

**Comment by DimitriPapadopoulos** [comment]
Indeed _openfortivpn_ uses GCD semaphores:
```
#ifdef __APPLE__
/* Mac OS X defines sem_init but actually does not implement them */
#include <dispatch/dispatch.h>

typedef dispatch_semaphore_t	sem_t;

#define sem_init(psem,x,val)	*psem = dispatch_semaphore_create(val)
#define sem_post(psem)		dispatch_semaphore_signal(*psem)
#define sem_wait(psem)		dispatch_semaphore_wait(*psem, \
					DISPATCH_TIME_FOREVER)

#define sem_destroy(psem)	dispatch_release(*psem)
#endif
```
And indeed using GCD semaphore in signal handlers is documented to be unsafe. I'm surprised there aren't more references in Google about that. Impressive investigation by the way! :thumbsup:

I haven't (ever?) used semaphores recently so perhaps you can help here:
- Are named POSIX semaphores an alternative? Named semaphores can be shared by several processes. Therefore they may outlive the process that created them and float around until the system is rebooted. Other downsides we should be aware of? Differences with the [pipe alternative](http://ck-hack.blogspot.no/2013/09/unnamed-semaphores-and-pososx.html)?
- Have you experienced actual problems with the current _openfortivpn_ code? I'm asking because designing alternative and portable code seems a little bit involved - especially since many regular_openfortivpn_ contributors might not own a Mac.
- BSD kqueue/kevent wouldn't work on Linux, would it? Should we use unnamed POSIX semaphores on Linux and BSD kqueue/kevent on Mac OS and BSDs? Other portable alternatives?

**Comment by henninglive** [comment]
I am not an expert on this, I just stumbled on the problem while trying to find a workaround for missing unnamed semaphores on OSX. I saw people suggesting GCD semaphores as an alternative and while researching if they were safe, I found your project. I have never actually tried *openfortivpn*, but here is my insight anyway.

* Using a pipe is probably and best easiest solution, there isn’t any downside as far as I am aware.
* Named semaphores would also work, but as you mentioned, you need take special care to clean them up. You would attempt to destroy them both on startup and on shutdown just in case. There is also the problem with short name limits, you need to be careful to avoid name conflicts.
* The BSD kqueue/kevent solution would not work on Linux, you would need to use POSIX semaphores. This is probably not what you want, the OpenJDK guy is suggesting it because kqueue/kevent is the preferred way to wait for signals on OSX. The API is much nicer then POSIX.

**Comment by mrbaseman** [comment]
I have found [this about Mach semaphores](https://books.google.de/books?id=K8vUkpOXhN4C&pg=PA1219&lpg=PA1219&dq=semaphore_create&source=bl&ots=OLqlSXWwWA&sig=k0WslTHLo1BevUXuh_iOSikpM0w&hl=de&sa=X&ved=0ahUKEwjtlP-7153UAhUDfFAKHQzwDIwQ6AEINjAJ) and I think the current approach in #135 looks promising. Do you agree?

As an ingredient for the more general discussion, what about unnamed [POSIX semaphores](https://linux.die.net/man/7/sem_overview) if they are initialized by a global variable?

**Comment by henninglive** [comment]
Ok, so POSIX named semaphores are actually implemented on top of Mach semaphores. Weird that the [documentation](https://developer.apple.com/library/content/documentation/Darwin/Conceptual/KernelProgramming/synchronization/synchronization.html) suggests they are not async signal safe, they might be safe, but I don’t know if I would take the chance. I would be very careful using anything to not explicitly guaranteeing async signal safety, there is so much that can go wrong with signal handlers.

I still think using a pipe is best idea. It would work with pthread_cancel() and the possible downsides(Performance and Unnecessary use of extra file descriptors) shouldn’t be a concern for you.

What do you mean by *initialized by a global variable*, are you talking about unnamed semaphores in memory shared between processes or are you talking about static initializer like `PTHREAD_MUTEX_INITIALIZER`.  I don’t think that exists for semaphores.

**Comment by mrbaseman** [comment]
I was referring to Unnamed semaphores in memory shared between threads 
> A thread-shared semaphore is placed in an area of memory shared between the threads of a process, for example, a global variable.

**Comment by henninglive** [comment]
Isn't that the normal use case?

**Comment by mrbaseman** [comment]
Probably yes, at least that  is  how they are used in openfortivpn

**Comment by mrbaseman** [comment]
We are still left with the question if semaphores on MacOS X are async-signal-safe. I have been further studying the various variants and I have found an interesting [handout](https://see.stanford.edu/materials/icsppcs107/23-Concurrency-Examples.pdf). 

> A semaphore is somewhat like an integer variable, but is special in that its operations
(increment and decrement) are guaranteed to be atomic—you cannot be halfway
through incrementing the semaphore and be interrupted and waylaid by another thread
trying to do the same thing.

#135 in its current version would switch from GCD semaphores to Mach semaphores. (named) POSIX semaphores seem to be [implemented upon Mach semaphores](https://books.google.de/books?id=K8vUkpOXhN4C&pg=PA1219&lpg=PA1210&focus=viewport&dq=semaphore_create&hl=de) but we don't know if extra synchronization efforts are involved. Anyhow, as @henninglive initially has pointed out, the [documentation](https://developer.apple.com/library/content/documentation/Darwin/Conceptual/KernelProgramming/synchronization/synchronization.html) states 

> Semaphores can be used any place where mutexes can occur. This precludes their use in interrupt handlers...

When looking for more detailed information about this I came across [this thread](http://www.linuxquestions.org/questions/programming-9/locking-mutex-in-a-signal-handler-function-789218/) saying that the problem could be deadlocks when a mutex is used in a signal handler. Although this is not exactly what we have in our case, let's look how semaphores are used in openfortivpn:

- The critical part is the signal handler, and indeed in current master `dispatch_semaphore_signal` is used there, and in # 135 this changes to `semaphore_signal`. GCD [avoids going down to the kernel when ressources are available](https://developer.apple.com/library/content/documentation/General/Conceptual/ConcurrencyProgrammingGuide/OperationQueues/OperationQueues.html#//apple_ref/doc/uid/TP40008091-CH102-SW24) whereas the direct calls to the Mach functions always operate on the semaphore. 
- Now, what does the [semaphore_signal}(http://web.mit.edu/darwin/src/modules/xnu/osfmk/man/semaphore_signal.html) actually do? 
> The semaphore_signal function increments the semaphore count. If the count goes non-negative (i.e. greater than or equal to 0) and a thread is blocked on the semaphore, then the waiting thread is scheduled to execute...
- what can basically go wrong, if we can believe the handout and the increment can not be interrupted and the blocked thread is just put back into the [queue for scheduling](https://developer.apple.com/library/content/documentation/Darwin/Conceptual/KernelProgramming/synchronization/synchronization.html#//apple_ref/doc/uid/TP30000905-CH218-TPXREF108)?
> Mach semaphores obey Mesa semantics—that is, when a thread is awakened by a semaphore becoming available, it is not executed immediately. 
- the main thread is waiting on `sem_stop_io` and any thread that goes into the signal handler just tries to wake up this main thread which then would continue to cancel all threads, join them and do all the other cleanup. There is no `semaphore_wait` or alike in the signal handler which would result in deadlocks, and if I understand it correctly, there shouldn't be a problem that several threads attempt to do the same thing

@henninglive by incidence I have found your thread [here](https://www.reddit.com/r/osx/comments/6exsqd/mach_semaphores_in_posix_signal_handlers/), so I'm linking this in case some new insight will be posted there by someone.

**Comment by henninglive** [comment]
[Here](https://github.com/st3fan/osx-10.9/blob/34e34a6a539b5a822cda4074e56a7ced9b57da71/xnu-2422.1.72/libsyscall/mach/semaphore.c) is the source for the userspace part of mach semaphores, straight syscalls. [Here](https://github.com/st3fan/osx-10.9/blob/34e34a6a539b5a822cda4074e56a7ced9b57da71/xnu-2422.1.72/bsd/kern/posix_sem.c#L937-L993) is the userspace part of the source for sem_post(), which we know is async signal safe on named semaphores. There is mutex here, might be relevant.

**Comment by mrbaseman** [comment]
@henninglive thank you very much for this analysis and for the references that you have provided. 
I have tried to do an analysis of GCD semaphores, too. The [documentation](https://developer.apple.com/library/content/documentation/General/Conceptual/ConcurrencyProgrammingGuide/OperationQueues/OperationQueues.html#//apple_ref/doc/uid/TP40008091-CH102-SW24) sais 

> A dispatch semaphore works like a regular semaphore with one exception. When resources are available, it takes less time to acquire a dispatch semaphore than it does to acquire a traditional system semaphore. This is because Grand Central Dispatch does not call down into the kernel for this particular case. The only time it calls down into the kernel is when the resource is not available and the system needs to park your thread until the semaphore is signaled.

However, I can't draw a similar comparison between the [implementation](https://github.com/st3fan/osx-10.9/blob/master/libdispatch-339.1.9/src/semaphore.c) and the one of POSIX named semaphores. So, given the fact that  POSIX named semaphores are actually implemented on top of Mach semaphores, moving over to Mach semaphores probably is a good step forward. 

@adrienverge and @DimitriPapadopoulos I'm now convinced that we should accept #135. It not only solves the locking issue that @Mabin-J has observed, but also addresses the possible issue with GCD semaphores discussed here. Do you still have any objections? Perhaps a very short summary of this profound discussion here in #105 should be included in the commit message, and perhaps we can prepare a wording here that we can then propose to @Mabin-J

**Comment by adrienverge** [comment]
@mrbaseman Looks good to me! :+1:

**Comment by DimitriPapadopoulos** [comment]
@mrbaseman Looks good to me too. Although I don't have time to test on Linux and cannot test on Mac, the whole discussion about different types of semaphores and their limitations is thorough and moving from GCD to Mach semaphores seems to make sense based on the available information.

**Comment by mrbaseman** [comment]
Ok, I have just tested it successfully and asked for an update of the commit message in https://github.com/adrienverge/openfortivpn/pull/135#issuecomment-307048969

**Comment by mrbaseman** [comment]
I have just merged #135, so I think we can also close this issue (see discussion above)
